src/kernel/chainparams.cpp

```diff
@@ -0,0 +1,510 @@
+        assert(genesis.hashMerkleRoot == uint256S("0x4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b"));
+
+        vFixedSeeds.clear();
+        vSeeds.clear();
+        // nodes with support for servicebits filtering should be at the top
+        vSeeds.emplace_back("testnet-seed.bitcoin.jonasschnelli.ch.");
+        vSeeds.emplace_back("seed.tbtc.petertodd.org.");
+        vSeeds.emplace_back("seed.testnet.bitcoin.sprovoost.nl.");
+        vSeeds.emplace_back("testnet-seed.bluematt.me."); // Just a static list of stable node(s), only supports x9
+
+        base58Prefixes[PUBKEY_ADDRESS] = std::vector<unsigned char>(1,111);
+        base58Prefixes[SCRIPT_ADDRESS] = std::vector<unsigned char>(1,196);
+        base58Prefixes[SECRET_KEY] =     std::vector<unsigned char>(1,239);
+        base58Prefixes[EXT_PUBLIC_KEY] = {0x04, 0x35, 0x87, 0xCF};
+        base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
+
+        bech32_hrp = "tb";
+
+        vFixedSeeds = std::vector<uint8_t>(std::begin(chainparams_seed_test), std::end(chainparams_seed_test));
+
+        fDefaultConsistencyChecks = false;
+        fRequireStandard = false;
+        m_is_test_chain = true;
+        m_is_mockable_chain = false;
+
+        checkpointData = {
+            {
+                {546, uint256S("000000002a936ca763904c3c35fce2f3556c559c0214345d31b1bcebf76acb70")},
+            }
+        };
+
+        m_assumeutxo_data = MapAssumeutxo{
+            // TODO to be specified in a future patch.
+        };
+
+        chainTxData = ChainTxData{
+            // Data from RPC: getchaintxstats 4096 0000000000000004877fa2d36316398528de4f347df2f8a96f76613a298ce060
+            .nTime    = 1661705221,
+            .nTxCount = 63531852,
+            .dTxRate  = 0.1079119341520164,
+        };
+    }
+};
+
+/**
+ * Signet: test network with an additional consensus parameter (see BIP325).
+ */
+class SigNetParams : public CChainParams {
+public:
+    explicit SigNetParams(const CChainParams::SigNetOptions& options)
+    {
+        vSeeds.clear();
+        vSeeds = options.seeds;
+
+        if (options.challenge.empty()) {
+            consensus.nMinimumChainWork = uint256S("0x000000000000000000000000000000000000000000000000000001291fc22898");
+            consensus.defaultAssumeValid = uint256S("0x000000d1a0e224fa4679d2fb2187ba55431c284fa1b74cbc8cfda866fd4d2c09"); // 105495
+            m_assumed_blockchain_size = 1;
+            m_assumed_chain_state_size = 0;
+            chainTxData = ChainTxData{
+                // Data from RPC: getchaintxstats 4096 000000d1a0e224fa4679d2fb2187ba55431c284fa1b74cbc8cfda866fd4d2c09
+                .nTime    = 1661702566,
+                .nTxCount = 1903567,
+                .dTxRate  = 0.02336701143027275,
+            };
+            consensus.signet_challenge = ParseHex("512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430210359ef5021964fe22d6f8e05b2463c9540ce96883fe3b278760f048f5189f2e6c452ae");
```

Comment:
```suggestion
            consensus.signet_challenge = ParseHex("512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430210359ef5021964fe22d6f8e05b2463c9540ce96883fe3b278760f048f5189f2e6c452ae");
            if (vSeeds.empty()) {
                vSeeds.push_back("seed.signet.bitcoin.sprovoost.nl.");
                vSeeds.push_back("178.128.221.177");
                vSeeds.push_back("v7ajjeirttkbnt32wpy3c6w3emwnfr3fkla7hpxcfokr3ysd3kqtzmqd.onion:38333");
            }
```

along with defaulting `options.seeds` to empty. Otherwise just overriding `options.challenge` will have you connecting to the default seeds which would be annoying.

Reply:
Thanks for review, I pushed another potential solution using constants. Let me know if you like something more akin to your solution, or fully using defaults better.

Reply:
Currently, if you run`bitcoind -signet -signetchallenge=51` then you start up with no seed nodes; with your current patch, you'll instead connect to the default signet seednodes, and then reject them for having the wrong network magic. That just seems buggy to me? Putting these defaults into the code rather than the structure definition seems like the easiest way of making that work sensibly to me.

Reply:
I thought I changed the behaviour now: It does not populate the seeds if the passed in option seeds are default and the signet challenge is non-default. This is still kind of confusing in the code though, so please let me know if you find that inferior.

Reply:
Err, somehow I missed that entirely? Whoops.

It's not the way I'd do it, but as long as it works, I don't really mind.

Another approach would be to have `SigNetOptions` default to an empty challenge and empty seed vector, but define a `const SigNetOptions GlobalDefaultSigNet { DEFAULT_CHALLENGE, DEFAULT_SEEDS };` with the global defaults, then make `ReadSigNetArgs` do:

```c++
if (args.IsArgSet("-signetchallenge")) {
    ...
    options.challenge = ParseHex(signet_challenge[0]);
} else {
    options = GlobalDefaultSigNet;
}
if (args.IsArgSet("-signetseednode")) {
    options.seeds = args.GetArgs("-signetseednode");
}
```

then you don't have to special case challenge/seeds in `SigNetParams(options)`, though you'd still be special casing min chain work, assume valid, etc.

Reply:
Commit "Decouple SigNetChainParams from ArgsManager" e893056ed68d02ef94f7e94fc6287217610723a4 now uses a struct with optionals as proposed here.

I tried implementing similar optionals for the RegTestOptions in commit "Douple RegTestChainParams from ArgsManager" 6d57a29df85277964edb21ee251b395e4eae1ef6, but it felt un-ergonomic.

Reply:
Using optionals for signet but not regtest makes sense to me.